### PR TITLE
Sphinx - Tatiana Snook - Wei Qiang

### PR DESCRIPTION
**Pull request by tatianasnook**

### CODE REVIEW

File: viewing_party/party.py [hunk 1]
```diff
@@ -1,23 +1,170 @@
...
 def create_movie(title, genre, rating):
-    pass
+    new_movie = {
+        "title": title,
+        "genre": genre,
+        "rating": rating
+    }
+    if title is None or genre is None or rating is None:
+        return None
+    return new_movie
```
Comment (mikellewade) [inline viewing_party/party.py:11]: So if we look at the directions in the README we see that we need to return  `None` if: 

If `title` is falsy, `genre` is falsy, or `rating` is falsy, this function should return `None`

Right now your function is only returning `None` if the values of the arguments of your function are `None` but what happens if you are one of the values is an empty string or `0`? How could we change your function to consider these questions?

File: viewing_party/party.py [hunk 2]
```diff
@@ -1,23 +1,170 @@
...
+
+def add_to_watchlist(user_data,movie):
+    user_data["watchlist"].append(movie)
+    return user_data
+
+
+def watch_movie(user_data,title):
+    for movies_to_watch in user_data["watchlist"]:
+        if title == movies_to_watch["title"]:
+            user_data["watched"].append(movies_to_watch)
```
Comment (mikellewade) [inline viewing_party/party.py:27]: Don't we have a function that does this for us? 👀

File: viewing_party/party.py [hunk 3]
```diff
@@ -1,23 +1,170 @@
...
+            user_data["watchlist"].remove(movies_to_watch)
```
Comment (mikellewade) [inline viewing_party/party.py:28]: So now that we have talked about Big O analysis, we know that `.remove` has a time complexity of `O(n)` since it needs to look through every element in a list to find the sought after value. Right now, this loop has an outer loop of  `O(n)` and you when you find the movie to remove you have another `O(n)` operation. How could we circumvent using `remove`? (Hint: Think about list comprehension or  ~~`pop` method~~) 

Also in this implementation, we will only use `remove` once, but in other filtering problems we might have to do this for multiple elements so it doesn't hurt to get in the habit now.

File: viewing_party/party.py [hunk 4]
```diff
@@ -1,23 +1,170 @@
...
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+    if len(user_data["watched"]) == 0:
+        return 0.0
+    sum_rating = 0.0
+    for movie in user_data["watched"]:
+        sum_rating += movie["rating"]   
+    avg_rating = float(sum_rating/len(user_data["watched"]))
```
Comment (mikellewade) [inline viewing_party/party.py:41]: The `/` operator in Python will implicitly result in a float, is there another reason why this is wrapped in the `float` constructor?

File: viewing_party/party.py [hunk 5]
```diff
@@ -1,23 +1,170 @@
...
+    for movie in user_data["watched"]:
+        count_genre[movie["genre"]] = count_genre.get(movie["genre"],0) + 1
+    
+    max_count = 0
+    popular_genre = None
+    for gerne,count in count_genre.items():
+        if count > max_count:
+            popular_genre = gerne
+            max_count = count   
+    return popular_genre
```
Comment (mikellewade) [inline viewing_party/party.py:58]: Great function! Love everything about it! ⭐️

File: viewing_party/party.py [hunk 6]
```diff
@@ -1,23 +1,170 @@
...
 # -----------------------------------------
 
-        
+def get_unique_watched(user_data):
+    user_unique_movie_list = [] 
+    friends_watched_titles = []
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_watched_titles.append(movie["title"])
```
Comment (mikellewade) [inline viewing_party/party.py:71]: Love how you are flattening the list here so you can simplify your following loop!

File: viewing_party/party.py [hunk 7]
```diff
@@ -1,23 +1,170 @@
...
+
+    for movie in user_data["watched"]:
+        if movie["title"] not in friends_watched_titles:
```
Comment (mikellewade) [inline viewing_party/party.py:74]: Right now, since `friends_watched_titles` is a list this line of code has an `O(n)` time complexity. What other data structure could we opt to use so we could circumvent the linear complexity and get one that is constant for this line?

File: viewing_party/party.py [hunk 8]
```diff
@@ -1,23 +1,170 @@
...
+    friends_unique_movies = []
+    user_watched_titles = []
+
+    for movie in user_data["watched"]:
+        user_watched_titles.append(movie["title"])
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie["title"] not in user_watched_titles \
+                and movie not in friends_unique_movies:
```
Comment (mikellewade) [inline viewing_party/party.py:90]: ⭐️

File: viewing_party/party.py [hunk 9]
```diff
@@ -1,23 +1,170 @@
...
+    user_watched_titles = get_user_watched_list(user_data)
+    
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie["title"] not in user_watched_titles \
+                and movie["host"] in user_data["subscriptions"] \
+                and movie not in recommended_movies_list:
+                    recommended_movies_list.append(movie)
+
+    return recommended_movies_list
```
Comment (mikellewade) [inline viewing_party/party.py:108]: Thinking about the requirements of if a movie can be added to the `recommend_movies_list`:

- Determine a list of recommended movies. A movie should be added to this list if and only if:
  - The movie is in the user's "favorites"
  - None of the user's friends have watched it

Don't we have a function that takes care of the second requirement? How could we change our function here so that it utilizes that function?

File: viewing_party/party.py [hunk 10]
```diff
@@ -1,23 +1,170 @@
...
+    fav_genre = get_favorite_genre(user_data)
+    user_watched_titles = get_user_watched_list(user_data)
+    friend_watched = get_friends_watched_list_unique(user_data)
+    
+    for movie in friend_watched:
+        if (movie["title"] not in user_watched_titles \
+            and movie["genre"] == fav_genre \
+            and movie not in recs):
+            recs.append(movie)
+    return recs
```
Comment (mikellewade) [inline viewing_party/party.py:129]: This function could be implemented without `user_watched_titles`, I encourage you to refactor to see how. It could make your code here a bit more concise.

File: viewing_party/party.py [hunk 11]
```diff
@@ -1,23 +1,170 @@
...
 
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
+def get_new_rec_by_genre(user_data):
+    recs = []
+    if not user_data["watched"]:
+        return recs   
+    if not any(friend["watched"] for friend in user_data["friends"]):
+        return recs  # Return empty list since there are no watched movies by friends
```
Comment (mikellewade) [inline viewing_party/party.py:118]: Usually with guard clauses, we want to check to if the data we are about to work with will cause the logic following the guard clause to result in an undefined value/error. Here the loops will still work as intended and return an empty list for the result. Love to see you experimenting with `any` method!

File: viewing_party/party.py [hunk 12]
```diff
@@ -1,23 +1,170 @@
...
+    genre_freq= {}
+    fav_genre = None
+    max_count = 0  
+    for movie in user_data["watched"]:   
+        genre = movie["genre"]
+        genre_freq[genre] = genre_freq.get(genre,0) + 1
+        if genre_freq[genre] > max_count:
+            max_count = genre_freq[genre]
+            fav_genre =genre
+    return fav_genre
```
Comment (mikellewade) [inline viewing_party/party.py:166]: Seems to me that this does the same thing as `get_most_watched_genre`.

File: viewing_party/party.py [hunk 13]
```diff
@@ -1,23 +1,170 @@
...
 
+# Assumed user_watched_list helper function
+def get_user_watched_list(user_data):
+    return [movie["title"] for movie in user_data["watched"]]
```
Comment (mikellewade) [inline viewing_party/party.py:170]: Love the helper functions! Now that we know about imports, could we move them to a different file and import them? 👀

File: viewing_party/party.py [hunk 14]
```diff
@@ -1,23 +1,170 @@
...
+    friend_watched = get_friends_watched_list_unique(user_data)
+    friends_watched_titles = []
+    
+    for movie in friend_watched:
+        friends_watched_titles.append(movie["title"])
+   
+    for movie in user_data["favorites"]:
+        if movie["title"] not in friends_watched_titles:
+                recs.append(movie)
+    return recs
```
Comment (mikellewade) [inline viewing_party/party.py:142]: ⭐️

File: tests/test_wave_05.py [hunk 15]
```diff
@@ -52,13 +52,18 @@ def test_new_genre_rec_from_empty_friends():
             }
         ]
     }
+     # Act
+    recommendations = get_new_rec_by_genre(sonyas_data)
+
+    assert len(recommendations) == 0
```
Comment (mikellewade) [inline tests/test_wave_05.py:58]: What if we did assert `len(genre_recommendations) == []`? This more readily declares that the genre_recommendations should be an empty list. Also for testing, if someone returned, say an empty string, this test would technically pass.

File: tests/test_wave_03.py [hunk 16]
```diff
@@ -43,24 +43,27 @@ def test_friends_unique_movies():
...
 
     # Act
     friends_unique_movies = get_friends_unique_watched(amandas_data)
 
     # Assert
     assert len(friends_unique_movies) == 3
-
-    raise Exception("Test needs to be completed.")
+    
+    # raise Exception("Test needs to be completed.")
```
Comment (mikellewade) [inline tests/test_wave_03.py:59]: Looks like this test wasn't completed, what could we assert here to make properly test this case?

File: tests/test_wave_01.py [hunk 17]
```diff
@@ -183,12 +186,16 @@ def test_moves_movie_from_watchlist_to_watched():
...
     assert len(updated_data["watched"]) == 2
     
-    raise Exception("Test needs to be completed.")
+    #raise Exception("Test needs to be completed.")
     # *******************************************************************************************
     # ****** Add assertions here to test that the correct movie was added to "watched" **********
     # *******************************************************************************************
+    assert updated_data["watched"][-1]["title"] == HORROR_1["title"]
+    assert updated_data["watched"][-1]["genre"] == HORROR_1["genre"]
+    assert updated_data["watched"][-1]["rating"] == HORROR_1["rating"]
```
Comment (mikellewade) [inline tests/test_wave_01.py:195]: ✅

This could also be condensed to:
```python
assert updated_data["watched"] == HORROR_1
```

File: tests/test_wave_01.py [hunk 18]
```diff
@@ -159,12 +159,15 @@ def test_moves_movie_from_watchlist_to_empty_watched():
...
-    raise Exception("Test needs to be completed.")
+    #raise Exception("Test needs to be completed.")
     # *******************************************************************************************
     # ****** Add assertions here to test that the correct movie was added to "watched" **********
     # *******************************************************************************************
-
-@pytest.mark.skip()
+    assert updated_data["watched"][-1]["title"] == MOVIE_TITLE_1
+    assert updated_data["watched"][-1]["genre"] == GENRE_1
+    assert updated_data["watched"][-1]["rating"] == RATING_1
```
Comment (mikellewade) [inline tests/test_wave_01.py:168]: Nice! This could look like this as well: 
```python 
assert updated_data["watched"][-1] == {
            "title": MOVIE_TITLE_1,
            "genre": GENRE_1,
            "rating": RATING_1
        }
```
